perceval/client.py

```diff
@@ -0,0 +1,213 @@
+#
+# Authors:
+#     Valerio Cosentino <valcos@bitergia.com>
+#
+
+import logging
+import time
+
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3.util import Retry
+
+from .errors import RateLimitError
+
+logger = logging.getLogger(__name__)
+
+
+class HttpClient:
+    """Abstract class for HTTP clients.
+
+    Base class to query data sources taking care of retrying requests in case connection issues. If the data source does
+    not send back a response after retrying a request, a RetryError exception is thrown.
+
+    Derivated classes can use the methods fetch and _fetch to obtain data from the data source.
+
+    To track which version of the client was used during the fetching
+    process, this class provides a `version` attribute that each client
+    may override.
+
+    :param base_url: base URL of the data source
+    :param max_retries: number of max retries to a data source before raising a RetryError exception
+    :param default_sleep_time: default time to sleep in case of connection problems
+    :param rate_limit_handler: object to handle rate limit issues
+    """
+    version = '0.1'
+
+    DEFAULT_SLEEP_TIME = 1
+    MAX_RETRIES = 5
+
+    MAX_RETRIES_ON_CONNECT = 5
+    MAX_RETRIES_ON_READ = 5
+    MAX_RETRIES_ON_REDIRECT = 5
+    MAX_RETRIES_ON_STATUS = 5
+
+    DEFAULT_RETRY_AFTER_STATUS_CODES = [503, 413, 429]
+
+    GET = "GET"
+    POST = "POST"
+
+    def __init__(self, base_url, max_retries=MAX_RETRIES, default_sleep_time=DEFAULT_SLEEP_TIME,
+                 rate_limit_handler=None):
+        self.base_url = base_url
+        self.session = self.create_http_session()
+        self.max_retries = max_retries
+        self.default_sleep_time = default_sleep_time
+        self.rate_limit_handler = rate_limit_handler
+
+    def __del__(self):
+        self.close_http_session()
+
+    def create_http_session(self, headers=None):
+        session = requests.Session()
+
+        if headers:
+            session.headers.update(headers)
+
+        return session
+
+    def close_http_session(self):
+        self.session.close()
+
+    def set_retry_after_status_codes(self, status_codes=None):
+        if status_codes:
+            Retry.RETRY_AFTER_STATUS_CODES = frozenset(status_codes)
+        else:
+            Retry.RETRY_AFTER_STATUS_CODES = self.DEFAULT_RETRY_AFTER_STATUS_CODES
+
+    def fetch(self, url, payload=None, headers=None, method=GET,
+              stream=False,
+              retry_on_status_codes=None,
+              retry_after_status_code=None):
+        yield self._fetch(url, payload, headers, method, stream,
+                          retry_on_status_codes,
+                          retry_after_status_code)
+
+    def _fetch(self, url, payload=None, headers=None, method=GET,
+               stream=False,
+               retry_on_status_codes=None,
+               retry_after_status_codes=None):
+
+        self.sleep_for_rate_limit()
```

Comment:
Instead og having this code here, I would do the next, taking advantage of multiple inheritance:
Let's take `GitHub` class the base case:

```
class GitHub(HttpClient, RateLimitHandler):

    def fetch(self, url, payload=None, headers=None, method=GET, retry_on_status_codes=None, retry_after_status_code=None):
        self.sleep_for_rate_limit()
        yield super().fetch(url, payload, headers, method, stream, retry_on_status_codes, retry_after_status_code)
        self.update_rate_limit(response)
```

Thus,  `sleep_for_rate_limit` and `update_rate_limit` can be safely removed from `HttpClient`. In addition, GitHub client `__init__` will need to call the `__init__` method of both parets.

https://stackoverflow.com/questions/26927571/multiple-inheritance-in-python3-with-different-signatures

Maybe we can do the same with pagination? What do you think?